File: Leetcode Challenge/March/C++/Maximum Frequency Stack.cpp
class FreqStack {
    //max-heap
    priority_queue<pair<int, pair<int, int>>> pq;
    unordered_map<int, int> freq; 
    int pos = 0;
public:
    void push(int x) {
        pq.push(make_pair(++freq[x], make_pair(++pos, x)));
    }
    
    int pop() {
        auto val = pq.top();
        pq.pop();
        int x = val.second.second;
        freq[x]--;
        return x;
    }
};

File: Leetcode Challenge/March/Java/Maximum Frequency Stack.java
class FreqStack {

    Map<Integer, Integer> map;
    PriorityQueue<Tuple> heap;
    int order;
    
    public FreqStack() {
        this.map = new HashMap<>();
        this.heap = new PriorityQueue<>();
        this.order = 0;
    }
    
    public void push(int val) {
        map.put(val, map.getOrDefault(val, 0) + 1);
        heap.add(new Tuple(val, map.get(val), ++order));
    }
    
    public int pop() {
        Tuple tuple = heap.poll();
        int val = tuple.val;
        map.put(val, map.get(val) - 1);
        return val;
    }
}

class Tuple implements Comparable<Tuple> {
    
    Integer val;
    Integer freq;
    Integer order;
    
    Tuple(Integer val, Integer freq, Integer order) {
        this.val = val;
        this.freq = freq;
        this.order = order;
    }
    
    public int compareTo(Tuple other){
        int freqCompare = other.freq.compareTo(this.freq);
        return freqCompare != 0 ? freqCompare : other.order.compareTo(this.order);
    }
}
